### PR TITLE
Add "Capture Location" task

### LINKED ISSUE #774: [Data collection] Support "capture location" data collection task
Capture "latitude, longitude, altitude in meters, and accuracy radius"

Show values in form. In the future we can visualize the location on the map and show calculated distance from feature location.

Steps:

- [ ] Add scaffolding and basic UI - #1861 
- [ ] Parse the task from Firebase
- [ ] Saving the data back to local and remote storage

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Fixes #774

 * Adds support for "capture_location" type task
 * Adds info card for all map type screens with lat/lng, accuracy and altitude when location lock is enabled
 * Adds info card for all map type screens with lat/lng at viewport centre when location lock is disabled

@gino-m  PTAL?

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
@gino-m This PR is ready for a review. PTAL

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/polygon/PolygonDrawingTaskFragment.kt [hunk 1]
```diff
@@ -48,13 +50,10 @@ class PolygonDrawingTaskFragment : Hilt_PolygonDrawingTaskFragment<PolygonDrawin
...
     val rowLayout = LinearLayout(requireContext()).apply { id = View.generateViewId() }
+    polygonDrawingMapFragment = PolygonDrawingMapFragment.newInstance(viewModel, map)
     parentFragmentManager
       .beginTransaction()
-      .add(
-        rowLayout.id,
-        PolygonDrawingMapFragment.newInstance(viewModel, map),
-        "Draw a polygon fragment"
-      )
+      .add(rowLayout.id, polygonDrawingMapFragment, "Draw a polygon fragment")
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/polygon/PolygonDrawingTaskFragment.kt:56]: Same here - better to use class name or to omit.

File: ground/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -144,7 +143,7 @@
   <string name="drop_a_pin">Drop a pin</string>
   <string name="drop_a_pin_tooltip_text">Drag your map until the center pin is on the desired location</string>
   <string name="draw_an_area">Draw an area</string>
-  <string name="accuracy">Accuracy</string>
+  <string name="current_location">Current Location</string>
```
Comment (gino-m) [inline ground/src/main/res/values/strings.xml:146]: Nit: Use Title case: "Current location"

**Review by gino-m — CHANGES_REQUESTED** [review]
That's a lot of work Shobhit, quite impressive! A few suggestions inline, thank you!

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/polygon/PolygonDrawingTaskFragment.kt (hunk 1 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/polygon/PolygonDrawingTaskFragment.kt]: Done

File: ground/src/main/res/values/strings.xml (hunk 2 above)
Comment (shobhitagarwal1612) [reply ground/src/main/res/values/strings.xml]: Done

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/location/CaptureLocationTaskFragment.kt [hunk 3]
```diff
@@ -41,33 +44,27 @@ class CaptureLocationTaskFragment :
...
-    taskBinding.lifecycleOwner = this
-    taskBinding.viewModel = viewModel
-    return taskBinding.root
+    val rowLayout = LinearLayout(requireContext()).apply { id = View.generateViewId() }
+    parentFragmentManager
+      .beginTransaction()
+      .add(
+        rowLayout.id,
+        CaptureLocationMapFragment.newInstance(viewModel, map),
+        "Drop a pin fragment"
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/location/CaptureLocationTaskFragment.kt]: Done

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionConverter.kt [hunk 4]
```diff
@@ -147,6 +148,18 @@ internal object SubmissionConverter {
...
+  private fun putCaptureLocationResponse(
+    taskId: String,
+    obj: Any,
+    responses: MutableMap<String, TaskData>
+  ) {
+    val map = obj as HashMap<String, *>
+    val result = LocationTaskDataConverter.fromFirestoreMap(map).getOrNull()
+    if (result != null) {
+      responses[taskId] = result
+    }
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionConverter.kt:158]: Done

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/schema/GeometryConverter.kt [hunk 5]
```diff
@@ -87,13 +94,20 @@ object GeometryConverter {
...
+        POINT_TYPE -> geoPointToPoint(coordinates as GeoPoint)
+        POLYGON_TYPE -> nestedIndexedMapToPolygon(coordinates as IndexedMap<IndexedMap<GeoPoint>>)
+        MULTI_POLYGON_TYPE ->
+          nestedIndexedMapToMultiPolygon(
+            coordinates as IndexedMap<IndexedMap<IndexedMap<GeoPoint>>>
+          )
+        else -> throw DataStoreException("Invalid geometry type '$type'")
+      }
+    } catch (e: ClassCastException) {
+      Timber.e(e, "Invalid geometry format: $type, $coordinates")
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firebase/schema/GeometryConverter.kt]: Done

File: ground/src/main/java/com/google/android/ground/ui/common/AbstractMapFragmentWithControls.kt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    return binding.root
+  }
+
+  private fun setCurrentLocationAsInfoCard(location: Location?) {
+    if (location == null) {
+      binding.infoCard.visibility = View.GONE
+    } else {
+      binding.cardTitle.setText(R.string.current_location)
+      binding.cardValue.text = LocationTaskData.fromLocation(location).getDetailsText()
+      binding.infoCard.visibility = View.VISIBLE
```
[the remark this answers is not in the document]
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/ui/common/AbstractMapFragmentWithControls.kt]: Making it generic would require adding accuracy and altitude to the `CameraPosition` object. Simplified the methods to reuse shared code.

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    df.roundingMode = RoundingMode.DOWN
+    return "${LatLngConverter.processCoordinates(geometry.coordinates)}\n" +
+      "Altitude: ${df.format(altitude)}m\n" +
+      "Accuracy: ${df.format(accuracy)}m"
+  }
+
+  override fun isEmpty(): Boolean = geometry == null
+
+  companion object {
+    fun fromLocation(location: Location?): LocationTaskData? {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt:47]: Perhaps more idiomatic would be implement this as an extension function:

```kt
fun Location.toTaskData() {
  val altitude = if (hasAltitude()) altitude else null
  val accuracy = if (hasAccuracy()) accuracy else null
  return LocationTaskData(
          Point(Coordinates(latitude, longitude)),
          altitude,
          accuracy?.toDouble()
}
```

And usagr:
```kt
location?.toTaskData()
```

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import com.google.android.ground.ui.datacollection.tasks.point.LatLngConverter
+import java.math.RoundingMode
+import java.text.DecimalFormat
+
+data class LocationTaskData(
+  val geometry: Geometry?,
+  val altitude: Double?, // in metres
+  val accuracy: Double? // in metres
+) : TaskData {
+  override fun getDetailsText(): String {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt:31]: Are we still using this method anywhere? If not maybe we should delete it, or at least not implement new instances?

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    if (geometry !is Point) {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt:32]: If geometry can only be a `Point`?, can we just make the constructor only accept `Point`?

**Review by gino-m — APPROVED** [review]
Sorry, a few suggestions missed first time around. Overall LGTM.

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt (hunk 8 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt:31]: Yes, it is being used for all tasks to display the user visible text.

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt (hunk 9 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt]: Done

File: ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt (hunk 7 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/model/submission/LocationTaskData.kt]: Thanks! Done

**Pull request merged**